### ISSUE TITLE
embedding constraints

### ISSUE THREAD

**Issue by dribnet** [lmcinnes/umap#432]
Does UMAP currently support specifying any embedding constraints?

In the past I've found it useful when doing dimensionality reduction to specify targets for a few key inputs. For example, if embedding a UMAP of RGB colors on to 2D one can make the output more predictable by specifying that the darkest input go to the left and the lightest to the right. Currently I can get the same effect by rotating and scaling the UMAP embedding after the fact (which can be automated), but a more direct and flexible way would be to introduce an ability in the API to pin specific inputs to target destinations in the embedding.

**Comment by lmcinnes** [comment]
I don't know any way of doing that with the current code, but I agree that it certainly makes a lot of sense. I'll add this as a feature request, and hopefully I can add it as a feature in the near future.

**Comment by dribnet** [comment]
Thanks Leland - encouraged that this idea is not DOA let me motivate this a bit more providing a sample use case, proposed API, and baseline implementation.

One use case is when there really is an approximate low dimensional embedding lurking in the higher dimensional data. Here one could attempt to make a more canonical view which could be compared across different datasets, etc. For example, suppose the data was approximately a 2D doughnut embedded in 3D space - in my motivating case I'll make this a (noisy) HSV colorwheel within an RGB cube:

```python
		h = np.random.random()
		s = np.random.uniform(0.5, 1.0)
		v = np.random.uniform(0.8, 0.9)
		data[n] = colorsys.hsv_to_rgb(h, s, v)
```

UMAP generally has no problem recovering the structure of this data:

![colors_orig_3](https://user-images.githubusercontent.com/945979/82740269-664c7580-9d9b-11ea-80a7-1067b144ff1b.png)

However, when running it multiple times with different seeds for generating the data the UMAP "view" of the colorwheel slice will be arbitrarily rotated, translated, and even flipped. To save space I'll animate results across 3 runs:

![orig3](https://user-images.githubusercontent.com/945979/82740361-2d60d080-9d9c-11ea-89a1-cdef89b20af7.gif)

One simple and flexible solution to constraining the embedding would be the ability to specify "pinned data" to UMAP.fit() which constrains certain inputs to be assigned destination locations in the embedded space. In my example I'm going to find the most extreme points on each axis (ie: the index locations for the reddest, greenest, and bluest values) and assign them target points in a simple lookup table:

```python
	pinned_data = [[max_red, [0, 3]], [max_green, [6, 10]], [max_blue, [10, 0]]]
```

So first here's a plot showing just these three pinned points:

![colors_targets](https://user-images.githubusercontent.com/945979/82740705-5fbffd00-9d9f-11ea-9ba4-07b57f834ab3.png)

And if we pass these three destinations to UMAP.fit() we can get an embedding is constrained to not move these three selected inputs:

```python
	u = fit.fit_transform(data, pinned_data=pinned_data)
```

The result is that even when we change the seed to generate three different random datasets, the results are much easier to compare against each other.

![pined3](https://user-images.githubusercontent.com/945979/82740871-d1e51180-9da0-11ea-8003-5cd3686c1c93.gif)

You could even imagine aggregating these results across multiple data sources by combining the embedded outputs or even just overlaying the graphs directly with some transparency.

---

Now my current "implementation" is super brain-dead. All I do is repeatedly force-move each pinned point in the embedded space each epoch before calling the `optimize_fn`.

```python
    for n in range(n_epochs):
        if pinned_data is not None:
            for p in pinned_data:
                head_embedding[p[0]] = p[1]
        optimize_fn(...
```

So this version really only works by taking some liberties with `n_neighbors`, `n_epochs` and cherry picking a bit. But it communicates the idea - so if we are not put off too much by my API decisions around this and are willing to push this further, I can take a shot making this more general by also having the `pinned_data` influence the initial embedding which I think would make this more general.

**Comment by lmcinnes** [comment]
Thanks for the explanation. Your approach is not so fr off from what I had in mind. I was actually thinking of making it a little more general and having a mask (boolean vector) of fixed points.  Thus in the internal optimization we would simply check if a point is fixed or not before updating/moving it. If done right this should be relatively cheap and not slow things down noticeably. This opens up other possibilities as well, such as making the transform a little cleaner (all the training data will simply be declared fixed).

The downside of this approach is that we would still need some level of better front-end API on top of that to facilitate your use case (since setting up an initialization and designated fixed points would be a little much to expect of end users in general).

I would certainly be interested in seeing a PR if you feel you have something semi-ready to go of course.

**Comment by dribnet** [comment]
Great to hear you thinking along similar lines. No, I don't have any better implementation - just was trying to communicate that use case more concretely.

There is another use case I come across frequently: grid layout. Here's a prototypical use case: flag layout by color:

![gridlab_sub3p](https://user-images.githubusercontent.com/945979/83041725-e83ae800-a094-11ea-892a-fa26fef26003.jpg)

[My current codebase](https://github.com/vusd/smartgrid) for this first generates a 2D umap scatterplot:

![tsne](https://user-images.githubusercontent.com/945979/83041834-03a5f300-a095-11ea-9499-595326d20bc3.png)

But then does a followup [Linear Assignment Problem solver](https://github.com/gatagat/lap) step to move points to the nearest grid location.

![movement](https://user-images.githubusercontent.com/945979/83042078-5384ba00-a095-11ea-9ee3-c0993baebafc.png)

Assuming umap doesn't currently have a means for doing discrete layouts like this, we could potentially discuss whether this use case would also be a possible form of an allowable embedding constraint.

**Comment by lmcinnes** [comment]
Embedding constraint-wise that's certainly a little more complicated to implement directly. It certainly isn't impossible, but would require a different approach. I'll certainly keep it in mind however.

**Comment by dribnet** [comment]
I might have a better proposal for how to implement this. Going back to my original toy problem: 

> embedding a UMAP of RGB colors on to 2D one can make the output more predictable by specifying that the darkest input go to the left and the lightest to the right

Perhaps this could be done with a custom `output_metric(a,b)` that - in the X direction only - combines the existing output metric (euclidean) with (i think) the dot product of (b-a) and a preferred cardinal direction in the original space (in this example the higher dimension RGB "brightness vector" goes from (0,0,0) to (1,1,1)). 

Just saying this out loud for feedback 😅 - if it seems loosely sane I'll try it out and report back.
